### PR TITLE
OGM-662

### PR DESCRIPTION
**Pull request by emmanuelbernard**
Probably can be merged but it's more to get feedback that this pull request is opened.
Known remaining tasks:
- update the documentation
- add support from the programmatic API

But worse case that can wait until after the next release.

### CODE REVIEW

File: core/src/main/java/org/hibernate/ogm/util/configurationreader/impl/SimplePropertyReaderContext.java [hunk 1]
```diff
@@ -96,6 +99,26 @@ else if ( configuredValue instanceof Number ) {
 		}
 	}
 
+	private Integer getAsInteger() {
```
Comment (gunnarmorling) [inline core/src/main/java/org/hibernate/ogm/util/configurationreader/impl/SimplePropertyReaderContext.java]: `getAsInt()` could be implemented on top of this:

    Integer value = getAsInteger();
    return value != null ? value : 0;

File: core/src/main/java/org/hibernate/ogm/cfg/OgmProperties.java [hunk 2]
```diff
@@ -50,12 +50,18 @@
 
 	/**
 	 * Property for setting the host name to connect to. Accepts {@code String}.
+	 * Accepts a comma separated list of host / ports.
```
Comment (gunnarmorling) [inline core/src/main/java/org/hibernate/ogm/cfg/OgmProperties.java]: I'd explicitly mention the "bracket style" for IPv6.

File: core/src/main/java/org/hibernate/ogm/cfg/impl/HostParser.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+			return Hosts.NO_HOST;
+		}
+		// for each element between commas
+		String[] splits = hostString.split( "," );
+		for (String rawSplit : splits) {
+			// remove whitespaces
+			String split = rawSplit.trim();
+			//
+			Matcher matcher = HOST_AND_PORT_PATTERN.matcher( split );
+			String host = null;
```
Comment (gunnarmorling) [inline core/src/main/java/org/hibernate/ogm/cfg/impl/HostParser.java]: These two vars are unused.

File: core/src/main/java/org/hibernate/ogm/cfg/spi/Hosts.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	public Hosts(List<String> hosts, List<Integer> ports) {
+		int length = hosts.size();
+		Contracts.assertTrue( length == ports.size(), "Not the same number of hosts and ports" );
+		this.hostsAndPorts = new ArrayList<HostAndPort>( length );
+		for ( int i = 0 ; i < length; i++ ) {
+			hostsAndPorts.add( new HostAndPort( hosts.get( i ), ports.get( i ) ) );
+		}
+	}
+
+	public boolean hasNoHost() {
```
Comment (gunnarmorling) [inline core/src/main/java/org/hibernate/ogm/cfg/spi/Hosts.java]: I'd avoid the negated method name, e.g. `isEmpty()`.

**Comment by gunnarmorling** [comment]
Looks good to me overall, added some minor comments.

> add support from the programmatic API

Agreed that this can wait. Btw. I am wondering about the usefulness of giving connection-related settings through the API. It's something you typically would like to change without re-compiling the application, so using properties for that seems better suited to me.

Speaking of it, currently (pre ORM 5) we allow for two approaches for plugging in programmatic API configurations:

* naming a `Configurator` impl class via a property
* through `OgmConfiguration#configureOptionsFor()`

With `Configuration` being gone at some point, only the former would be possible. With the new bootstrap API, we can plug in an OGM-specific `SessionFactoryBuilder` which can be accessed through `unwrap()` and allows for OGM-specific bootstrap APIs. But that'd be "too late" for connection related settings, as `DatastoreProvider` (which is the client of those) isn't an SF-level service. It would work though for "runtime" settings, e.g. to control certain aspects of mapping.

**Comment by emmanuelbernard** [comment]
> Speaking of it, currently (pre ORM 5) we allow for two approaches for plugging in programmatic API configurations:
> 
> naming a Configurator impl class via a property
> through OgmConfiguration#configureOptionsFor()
> With Configuration being gone at some point, only the former would be possible. With the new bootstrap API, we can plug in an OGM-specific SessionFactoryBuilder which can be accessed through unwrap() and allows for OGM-specific bootstrap APIs. But that'd be "too late" for connection related settings, as DatastoreProvider (which is the client of those) isn't an SF-level service. It would work though for "runtime" settings, e.g. to control certain aspects of mapping.

Wouldn't it be better to offer it as a Hibernate `Service` and let the user plug an initiator in its service registry? I am not sure how that would look like but that would be at the right time and more natural I guess. thoughts?

**Comment by emmanuelbernard** [comment]
@gunnarmorling I've update things with your remarks and added a commit with the documentation.
Ready to go on my side.

**Comment by gunnarmorling** [comment]
> Wouldn't it be better to offer it as a Hibernate Service and let the user plug an initiator in its service registry?

Hm, maybe. Although the service/initiator stuff seems a bit tough on users. A configurator class made known through some prop still seems nicer to me.

**Comment by gunnarmorling** [comment]
Merged, thanks.

**Pull request merged**